Wrappers/Python/cil/optimisation/functions/Function.py

```diff
@@ -23,161 +23,204 @@
 import numpy as np
 from functools import reduce
 
+
 class Function(object):
-    
+
     r""" Abstract class representing a function 
-    
+
         Parameters
         ----------
-        
+
         L: number, positive, default None
             Lipschitz constant of the gradient of the function F(x), when it is differentiable.
-                
+
         Note
         -----
         The Lipschitz of the gradient of the function is a positive real number, such that :math:`\|f'(x) - f'(y)\| \leq L \|x-y\|`, assuming :math:`f: IG \rightarrow \mathbb{R}`
 
     """
-    
-    
-    def __init__(self, L = None):
+
+    def __init__(self, L=None):
         # overrides the type check to allow None as initial value
         self._L = L
-        
-    def __call__(self,x):    
-        
+
+    def __call__(self, x):
+
         raise NotImplementedError
 
     def gradient(self, x, out=None):
-        
         r"""Returns the value of the gradient of function :math:`F` at :math:`x`, if it is differentiable
-        
+
         .. math:: F'(x)
-        
+
+        Parameters
+        ----------
+        x : DataContainer
+
+        out: return DataContainer, if None a new DataContainer is returned, default None.
+
+        Returns
+        --------
+        DataContainer, the value of the gradient of the function at x. 
+
         """
         raise NotImplementedError
 
     def proximal(self, x, tau, out=None):
-        
         r"""Returns the proximal operator of function :math:`\tau F` at x        
-        
+
         .. math:: \text{prox}_{\tau F}(x) = \underset{z}{\text{argmin}} \frac{1}{2}\|z - x\|^{2} + \tau F(z)
-        
+
+        Parameters
+        ----------
+        x : DataContainer
+
+        tau: scalar
+
+        out: return DataContainer, if None a new DataContainer is returned, default None.
+
+        Returns
+        -------
+        DataContainer, the proximal operator of the function at x with scalar :math:`\tau`. 
+
         """
         raise NotImplementedError
 
     def convex_conjugate(self, x):
         r""" Returns the convex conjugate of function :math:`F` at :math:`x^{*}`,
```

Comment:
I now understand that this is the "evaluation of the function `F*` at `x`, where `F*` is the convex conjugate of function `F`"

Reply:
Thanks - I think it makes sense now